text/0000-official-utility-buildpacks.md

```diff
@@ -0,0 +1,75 @@
+
+# Definitions
+[definitions]: #definitions
+
+- **utility buildpack** - a buildpack that provides a generic capability that is useful across many vendors, platforms, and apps
+- **officially supported** - supported and maintained by the the Cloud Native Buildpacks project and its members
+
+# Motivation
+[motivation]: #motivation
+
+Many capabilities of lifecycle and pack do not require a native implementation within these components. These capabilities can (and have been) implemented at the user layer as buildpacks (ex. [inline-buildpack](https://registry.buildpacks.io/buildpacks/hone/inline)). In [PR #168](https://github.com/buildpacks/rfcs/pull/168) we assert that supporting these capabilities natively adds complexity to the implementation of lifecycle, which makes it difficult to innovate and improve.
+
+However, defering to the community and vendors to implement these kinds of utility buildpacks has not necessarily had good results. For example, there are three implementations of the [Procfile buildpack](https://registry.buildpacks.io/buildpacks/paketo-buildpacks/procfile). None of these implementations add significant (if any) advantages over the other. Yet buildpack users are force to decide which one to select. This is an unneccessary burden for a capability that could be provided by default.
+
+We seek to reduce complexity in the lifecycle and provide well maintained utility buildpacks for users by supporting them as official components of the Cloud Native Buildpacks project. This also aligns with our "everything's a buildpack" mantra.
+
+# What it is
+[what-it-is]: #what-it-is
+
+Reserve the `cnb` namespace for officially supported buildpacks. Then, on a case by case basis, we adopt utility buildpacks under this
+
+Examples of potential utility buildpacks include:
+
+- `cnb/profile` - Evaluating `.profile` scripts in the app repo during launch
+- `cnb/inline` - Executing [Inline buildpacks](https://github.com/buildpacks/rfcs/blob/main/text/0048-inline-buildpack.md)
+- `cnb/processes` - Setting app specific process types from a configuration file like `project.toml`
+- `cnb/env` - Setting customer environment variables defined in a configuration file like `project.toml`
+- `cnb/purge` - Purging files from the app repo before export (similar to [`.slugignore`](https://devcenter.heroku.com/articles/slug-compiler#ignoring-files-with-slugignore))
+
+# How it Works
+[how-it-works]: #how-it-works
+
+The utility buildpacks will be owned, supported, and maintained by the [Implementation team](https://github.com/buildpacks/community/blob/main/TEAMS.md#implementation-team).
+
+# Drawbacks
+[drawbacks]: #drawbacks
+
+- Bubbling these capabilities up to the user layer (as buildpacks) may add cognitive burden for users and errode the DX. We can mitigate this with new features like "default buildpacks".
+- New components to support
+- We have to ensure that each utility buildpack is portable and works in conjunction with all stacks
```

Comment:
Also all buildpack APIs...

Reply:
On the plus side this means we'll have to dogfood our breaking changes/migrations as a project.

Reply:
I think it is okay for the utility buildpacks to implement a single buildpack API?

Reply:
Would we have multiple versions of each utility buildpack (one per buildpack API)?

Reply:
I think it would make sense to bump the utility buildpacks API compat and cut a release each time the lifecycle adds a new api version. I also think that makes a good case for Implementation team owning these buildpacks.

Reply:
I'll be honest, I'm a bit wary of the implementation team owning these (we'd be using Buildpack Author Tooling to maintain them...) but at the same time, I could see the argument for it. We could use the utility buildpacks as test fixtures when adding support for a new api. They should ship together (or at least, concurrently). That said, coordination between the two teams to enable this should be possible!

Reply:
> I think it would make sense to bump the utility buildpacks API compat and cut a release each time the lifecycle adds a new api version. I also think that makes a good case for Implementation team owning these buildpacks.

By the same logic the implementation team should maintain `pack` as well, given `pack` should also upgrade to use the latest platform API after the lifecycle cuts a new release. Maybe we need to rename the implementation team because 98% of the code in the project is an implementation of at least one of the spec :p

Reply:
@ekcasey i think we should rename the team. :)

> They should ship together (or at least, concurrently). That said, coordination between the two teams to enable this should be possible!

@natalieparellano I feel like some joint maintenance makes sense but there isn't really that structure today in the project.